### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by bregman-arie**
In order to better understand the code and possibly have
a documentation generated based on the type annotations

### CODE REVIEW

File: cibyl/cli/output.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ class OutputStyle(Enum):
     """A machine-readable text based format."""
 
     @staticmethod
-    def from_key(key):
+    def from_key(key: object) -> 'OutputStyle':
```
Comment (jsanemet) [inline cibyl/cli/output.py:30]: I have already documented the parameter types in the docstring, there is no need to do it again. Having this information twice will only get us inconsistency between the two and more maintenance work.

File: cibyl/cli/argument.py [hunk 2]
```diff
@@ -64,11 +64,11 @@ def parse_ranges(self, expressions):
             parsed_expressions.append(Range(operator, operand))
         return parsed_expressions
 
-    def __bool__(self):
+    def __bool__(self) -> bool:
         return bool(self.value)
 
-    def __eq__(self, other):
+    def __eq__(self, other: 'Argument') -> bool:
```
Comment (jsanemet) [inline cibyl/cli/argument.py:70]: There is no need to type hint builtin functions. These are standard and well known enough that people already know what they are dealing with. Besides, hints will not help much as these are rarely called explicitly. For example, equality is called from a statement like: A == B and Python will not care about their types in such case.

File: cibyl/cli/validator.py [hunk 3]
```diff
@@ -83,7 +84,7 @@ def _consistent_system(self, system):
 
         return True
 
-    def _system_has_valid_sources(self, system):
+    def _system_has_valid_sources(self, system: object) -> bool:
```
Comment (jsanemet) [inline cibyl/cli/validator.py:87]: This is what I mean when talking about inconsistency, the docstring and the type hint are hinting for two different types here. The docstring one is more precise as is the one I would go with. So, it is a matter of having one or the other, but not both.

**Review by jsanemet — CHANGES_REQUESTED** [review]

File: cibyl/cli/output.py (hunk 1 above)
Comment (bregman-arie) [reply cibyl/cli/output.py:30]: If I remove it I get "cibyl/cli/output.py:30: error: Function is missing a type annotation" from mypy

File: cibyl/cli/validator.py (hunk 3 above)
Comment (bregman-arie) [reply cibyl/cli/validator.py:87]: It feels like we are re-discussing the whole idea of type hinting while we already did it

**Comment by jsanemet** [comment]
I am not against documenting our code better, but my fear is that we begin doing it blindly and without care. I have added a couple of comments indicating that some of these changes are way too much and that there is really no need to go so far in some cases. Type hints should be a tool for us to make collaboration between each other easier and to make bugs pop up sooner. I do not want it to be a source of unnecessary work.

**Comment by bregman-arie** [comment]
> I am not against documenting our code better, but my fear is that we begin doing it blindly and without care. I have added a couple of comments indicating that some of these changes are way too much and that there is really no need to go so far in some cases. Type hints should be a tool for us to make collaboration between each other easier and to make bugs pop up sooner. I do not want it to be a source of unnecessary work.

While I understand it to some extent I admit it conflicts to me with the decision we made on enforcing type hinting in CI. How do you see it works exactly?

**Comment by jsanemet** [comment]
I would even argue that the only place I would require them is at the interface between a component and the app's component. For example, if I would work with a component that implements a calculator, I would like for that Calculator class to be well documented, but I would not care as much for any class that sits below it.

**Comment by jsanemet** [comment]
Enforcing it as a must in the project is, at this point, too much. I have already documented all my code using docstrings and sphinx instead of type hints. Why would you like to repeat all that work again? It is already there.

Also, Python is not a strongly typed language. Myself, I have been trying to use it closer to how I would do Java, but OOP on this language is not as strong as in that one. That is because Python does not follow the same rules as Java and that is OK. Instead, it promotes the use of duck typing (the: hasattr / getattr we have used sometimes in the code). Type hinting is a tool to make typing stronger and there are parts in our codebase where it will not make sense.

**Comment by bregman-arie** [comment]
> Enforcing it as a must in the project is, at this point, too much. I have already documented all my code using docstrings and sphinx instead of type hints. Why would you like to repeat all that work again? It is already there.
> 
> Also, Python is not a strongly typed language. Myself, I have been trying to use it closer to how I would do Java, but OOP on this language is not as strong as in that one. That is because Python does not follow the same rules as Java and that is OK. Instead, it promotes the use of duck typing (the: hasattr / getattr we have used sometimes in the code). Type hinting is a tool to make typing stronger and there are parts in our codebase where it will not make sense.

Again, I can understand the argument to some extent, it's definitely debatable, but at the same time, I'm not following why this wasn't mentioned when we had a team discussion on it, and we decided to use it.

Anyway, I respect your opinion so let's put it on hold. Re-discuss it with the team and then we'll see how to continue.

File: cibyl/models/attribute.py [hunk 4]
```diff
@@ -22,26 +22,28 @@ class AttributeValue:
...
         self.attr_type = attr_type
         self.value = value
         self.arguments = arguments
 
-    def __bool__(self):
+    def __bool__(self) -> bool:
         return bool(self.value)
 
-    def __eq__(self, other):
+    def __eq__(self, other: 'AttributeValue') -> bool:
```
[the remark this answers is not in the document]
Comment (bregman-arie) [reply cibyl/models/attribute.py:35]: "having two types with the same name" - do you have an example?

[the remark this answers is not in the document]
Comment (jsanemet) [reply cibyl/models/attribute.py:35]: Yes, as an example there are two Job classes with the same name, one for Jenkins and one for Zuul. Their difference being in the package they are at.

**Comment by cescgina** [comment]
My 2 cents on this issue, that I already mentioned in another PR, where we had some discussion on type hints:
>I'd like to point out that type hints and docstrings, while they have some overlap, are not really equivalent. Docstrings are documentation (the types in the docstrings can be used by sphinx to generate a browsable package description) while the type hints enforced with mypy can be useful to detect bugs that in statically typed languages would be caught by the compiler.

That said, I think the best course of action would be to have this discussion again, making sure that any concerns are raised.

**Comment by bregman-arie** [comment]
Dropping it per team discussion

**Pull request closed without merging**